### PR TITLE
Dedupe all Make PVC API object functions into the one MakePersistentVolumeClaim to rule them all

### PR DESCRIPTION
**Pull request by davidz627**
We have a bunch of different create PVC functions. Dededupe them. 

/assign @msau42 
/cc @misterikkit @mkimuram

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by davidz627** [comment]
removed 2 `getClaim` functions and one `newClaim` function. Also moved around `volumeSource` creation for `base.go`

File: test/e2e/framework/pv_util.go [hunk 1]
```diff
@@ -621,13 +623,28 @@ func MakePersistentVolumeClaim(cfg PersistentVolumeClaimConfig, ns string) *v1.P
...
+		cfg.ClaimSize = "2Gi"
+	}
+
+	if len(cfg.NamePrefix) == 0 {
+		e2elog.Logf("Making PVC: NamePrefix unspecified, default: pvc-")
+		cfg.NamePrefix = "pvc-"
+	}
+
+	if cfg.VolumeMode != nil && *cfg.VolumeMode == "" {
+		e2elog.Logf("Warning: Making PVC: VolumeMode specified as invalid empty string, treating as nil")
```
Comment (davidz627) [inline test/e2e/framework/pv_util.go]: I know this isn't ideal but I wanted to keep the refactor small-ish. The issue is actually in the callers not using a pointer to the `VolumeMode` throughout the test code so they're defaulting to empty string sometimes. Would be a much larger refactor to fix - we can just treat them as nil here (to default) and log a warning

Comment (davidz627) [reply test/e2e/framework/pv_util.go]: fyi an empty string `VolumeMode` is actually an invalid value

File: test/e2e/storage/volume_provisioning.go [hunk 2]
```diff
@@ -1013,36 +1054,6 @@ func updateDefaultStorageClass(c clientset.Interface, scName string, defaultStr
 	verifyDefaultStorageClass(c, scName, expectedDefault)
 }
 
-func getClaim(claimSize string, ns string) *v1.PersistentVolumeClaim {
```
Comment (misterikkit) [inline test/e2e/storage/volume_provisioning.go]: these two were literally adjacent to each other in the same file? Golden.

**Review by misterikkit — COMMENTED** [review]
Just my thoughts on some of these calls, although I've probably misunderstood some of how these tests work.

File: test/e2e/framework/pv_util.go [hunk 3]
```diff
@@ -621,13 +623,28 @@ func MakePersistentVolumeClaim(cfg PersistentVolumeClaimConfig, ns string) *v1.P
 	// Specs are expected to match this test's PersistentVolume
```
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/framework/pv_util.go]: unclear. I think it doesn't actually mean anything. Looks like its referring to something that doesn't exist in the function anymore

File: test/e2e/framework/pv_util.go [hunk 4]
```diff
@@ -621,13 +623,28 @@ func MakePersistentVolumeClaim(cfg PersistentVolumeClaimConfig, ns string) *v1.P
...
+	if cfg.VolumeMode != nil && *cfg.VolumeMode == "" {
+		e2elog.Logf("Warning: Making PVC: VolumeMode specified as invalid empty string, treating as nil")
+		cfg.VolumeMode = nil
+	}
+
 	return &v1.PersistentVolumeClaim{
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: "pvc-",
+			GenerateName: cfg.NamePrefix,
 			Namespace:    ns,
```
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/framework/pv_util.go]: I believe namespace should also be in the config but I wasn't really sure so to keep this refactor simpler I decided not to change that. (main goal is to dedupe not to fix config)

File: test/e2e/framework/pv_util.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/framework/pv_util.go]: left as an exercise to the next person who comes along

File: test/e2e/framework/pv_util.go [hunk 5]
```diff
@@ -621,13 +623,28 @@ func MakePersistentVolumeClaim(cfg PersistentVolumeClaimConfig, ns string) *v1.P
 	// Specs are expected to match this test's PersistentVolume
 
 	if len(cfg.AccessModes) == 0 {
-		e2elog.Logf("AccessModes unspecified, default: all modes (RWO, RWX, ROX).")
+		e2elog.Logf("Making PVC: AccessModes unspecified, default: all modes (RWO, RWX, ROX).")
 		cfg.AccessModes = append(cfg.AccessModes, v1.ReadWriteOnce, v1.ReadOnlyMany, v1.ReadOnlyMany)
 	}
 
+	if len(cfg.ClaimSize) == 0 {
+		e2elog.Logf("Making PVC: ClaimSize unspecified, default: 2Gi")
```
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/framework/pv_util.go]: defaulting captured in PVCConfig comments

File: test/e2e/storage/csi_volumes.go [hunk 6]
```diff
@@ -119,8 +119,11 @@ func testTopologyPositive(cs clientset.Interface, suffix, namespace string, dela
 		addSingleCSIZoneAllowedTopologyToStorageClass(cs, class, topoZone)
 	}
 	test.Client = cs
-	test.Claim = newClaim(test, namespace, suffix)
-	test.Claim.Spec.StorageClassName = &class.Name
+	test.Claim = framework.MakePersistentVolumeClaim(framework.PersistentVolumeClaimConfig{
+		ClaimSize:        test.ClaimSize,
+		StorageClassName: &(class.Name),
+		VolumeMode:       &test.VolumeMode,
```
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/storage/csi_volumes.go]: I don't think its new behavior, one of the helper functions that took `StorageClassTest` as input was being used here and would set the volume mode based on this param

File: test/e2e/storage/mounted_volume_resize.go [hunk 7]
```diff
@@ -85,8 +85,11 @@ var _ = utils.SIGDescribe("Mounted volume expand", func() {
 		framework.ExpectNoError(err, "Error creating resizable storage class")
 		gomega.Expect(*resizableSc.AllowVolumeExpansion).To(gomega.BeTrue())
 
-		pvc = newClaim(test, ns, "default")
-		pvc.Spec.StorageClassName = &resizableSc.Name
+		pvc = framework.MakePersistentVolumeClaim(framework.PersistentVolumeClaimConfig{
+			ClaimSize:        test.ClaimSize,
```
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/storage/mounted_volume_resize.go]: we are in all instances where `StorageClassTest` is used

File: test/e2e/storage/regional_pd.go [hunk 8]
```diff
@@ -170,9 +173,12 @@ func testZonalFailover(c clientset.Interface, ns string) {
 		ExpectedSize: repdMinSize,
 	}
 	class := newStorageClass(testSpec, ns, "" /* suffix */)
-	claimTemplate := newClaim(testSpec, ns, "" /* suffix */)
-	claimTemplate.Name = pvcName
-	claimTemplate.Spec.StorageClassName = &class.Name
+	claimTemplate := framework.MakePersistentVolumeClaim(framework.PersistentVolumeClaimConfig{
+		NamePrefix:       pvcName,
+		ClaimSize:        testSpec.ClaimSize,
+		StorageClassName: &(class.Name),
```
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/storage/regional_pd.go]: ack. style preference :)

File: test/e2e/storage/volume_provisioning.go (hunk 2 above)
Comment (davidz627) [reply test/e2e/storage/volume_provisioning.go]: fun times

File: test/e2e/framework/pv_util.go (hunk 1 above)
Comment (davidz627) [reply test/e2e/framework/pv_util.go]: Yeah, a fix for another time

File: test/e2e/framework/pv_util.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidz627) [reply test/e2e/framework/pv_util.go]: removed new logs except the `VolumeMode` warning, documented in the `PVCConfig` type

**Comment by davidz627** [comment]
@misterikkit @msau42 
all comments addressed PTAL

File: test/e2e/storage/pvc_protection.go [hunk 9]
```diff
@@ -47,12 +47,16 @@ var _ = utils.SIGDescribe("PVC Protection", func() {
 		framework.ExpectNoError(framework.WaitForAllNodesSchedulable(client, framework.TestContext.NodeSchedulableTimeout))
 
 		ginkgo.By("Creating a PVC")
-		suffix := "pvc-protection"
+		prefix := "pvc-protection"
 		framework.SkipIfNoDefaultStorageClass(client)
-		testStorageClass := testsuites.StorageClassTest{
+		t := testsuites.StorageClassTest{
```
Comment (misterikkit) [inline test/e2e/storage/pvc_protection.go]: This test suite object does not add a lot of value to this test. I don't think it needs fixing, but this seems like a case where the values could have been inlined.

File: test/e2e/storage/testsuites/base.go [hunk 10]
```diff
@@ -241,6 +242,16 @@ func createGenericVolumeTestResource(driver TestDriver, config *PerTestConfig, p
 	return &r
 }
 
+func createVolumeSource(pvcName string, readOnly bool) *v1.VolumeSource {
```
Comment (misterikkit) [inline test/e2e/storage/testsuites/base.go]: I guess from a storage perspective, it made sense to split this one out? If your only goal was to de-dupe the instantiation of the PVC VolumeSource, then I think it was small enough to have twice.

**Review by misterikkit — COMMENTED** [review]
/lgtm

**Comment by davidz627** [comment]
/assign @krzyzacy 
for test framework approval

File: test/e2e/framework/pv_util.go [hunk 11]
```diff
@@ -613,21 +620,29 @@ func MakePersistentVolume(pvConfig PersistentVolumeConfig) *v1.PersistentVolume
...
 func MakePersistentVolumeClaim(cfg PersistentVolumeClaimConfig, ns string) *v1.PersistentVolumeClaim {
-	// Specs are expected to match this test's PersistentVolume
 
 	if len(cfg.AccessModes) == 0 {
-		e2elog.Logf("AccessModes unspecified, default: ReadWriteOnce (RWO).")
 		cfg.AccessModes = append(cfg.AccessModes, v1.ReadWriteOnce)
 	}
 
+	if len(cfg.ClaimSize) == 0 {
+		cfg.ClaimSize = "2Gi"
```
Comment (krzyzacy) [inline test/e2e/framework/pv_util.go]: is this documented anywhere? seems a bit magical to me

Comment (davidz627) [reply test/e2e/framework/pv_util.go]: Documented here: https://github.com/kubernetes/kubernetes/pull/80382/files#diff-420013911dac4a3b0f16035b6e90a571R96

Less magical than the previous defaulting even :)
There was logging everytime we defaulted that I removed: https://github.com/kubernetes/kubernetes/pull/80382#discussion_r305955175

**Review by krzyzacy — APPROVED** [review]

File: test/e2e/framework/pv_util.go (hunk 11 above)
Comment (krzyzacy) [reply test/e2e/framework/pv_util.go]: :+1:

**Pull request merged**